src/comms/Broker.cc

```diff
@@ -0,0 +1,224 @@
+    elem->Get<std::string>("unbind_service", this->dataPtr->unbindSrv).first;
+}
+
+//////////////////////////////////////////////////
+void Broker::Start()
+{
+  // Advertise the service for binding addresses.
+  if (!this->dataPtr->node.Advertise(this->dataPtr->bindSrv,
+                                     &Broker::OnBind, this))
+  {
+    ignerr << "Error advertising srv [" << this->dataPtr->bindSrv << "]"
+           << std::endl;
+    return;
+  }
+
+  // Advertise the service for unbinding addresses.
+  if (!this->dataPtr->node.Advertise(this->dataPtr->unbindSrv,
+                                     &Broker::OnUnbind, this))
+  {
+    ignerr << "Error advertising srv [" << this->dataPtr->unbindSrv << "]"
+           << std::endl;
+    return;
+  }
+
+  // Advertise the topic for receiving data messages.
+  if (!this->dataPtr->node.Subscribe(this->dataPtr->msgTopic,
+                                     &Broker::OnMsg, this))
+  {
+    ignerr << "Error subscribing to topic [" << this->dataPtr->msgTopic << "]"
+           << std::endl;
+    return;
+  }
+
+  igndbg << "Broker services:" << std::endl;
+  igndbg << "  Bind: [" << this->dataPtr->bindSrv << "]" << std::endl;
+  igndbg << "  Unbind: [" << this->dataPtr->unbindSrv << "]" << std::endl;
+  igndbg << "Broker topics:" << std::endl;
+  igndbg << "  Incoming messages: [" << this->dataPtr->msgTopic << "]"
+         << std::endl;
+}
+
+//////////////////////////////////////////////////
+std::chrono::steady_clock::duration Broker::Time() const
+{
+  return this->dataPtr->time;
+}
+
+//////////////////////////////////////////////////
+void Broker::SetTime(const std::chrono::steady_clock::duration &_time)
+{
+  this->dataPtr->time = _time;
+}
+
+//////////////////////////////////////////////////
+bool Broker::OnBind(const ignition::msgs::StringMsg_V &_req,
+                    ignition::msgs::Boolean &/*_rep*/)
+{
+  auto count = _req.data_size();
+  if (count != 3)
+  {
+    ignerr << "Receive incorrect number of arguments. "
+           << "Expecting 3 and receive " << count << std::endl;
+    return false;
+  }
+
+  std::string address = _req.data(0);
+  std::string model   = _req.data(1);
+  std::string topic   = _req.data(2);
+
+  std::lock_guard<std::mutex> lk(this->dataPtr->mutex);
+
+  if (!this->DataManager().AddSubscriber(address, model, topic))
+    return false;
+
+  ignmsg << "Address [" << address << "] bound to model [" << model
+         << "] on topic [" << topic << "]" << std::endl;
+
+  return true;
+}
+
+//////////////////////////////////////////////////
+void Broker::OnUnbind(const ignition::msgs::StringMsg_V &_req)
+{
+  auto count = _req.data_size();
+  if (count != 2)
+  {
+    ignerr << "Receive incorrect number of arguments. "
+           << "Expecting 2 and receive " << count << std::endl;
+    return;
+  }
+
+  std::string address = _req.data(0);
+  std::string topic   = _req.data(1);
+
+  std::lock_guard<std::mutex> lk(this->dataPtr->mutex);
```

Comment:
```suggestion
  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
```

Reply:
And more further...

Reply:
Renamed in [1140b56](https://github.com/ignitionrobotics/ign-gazebo/pull/1416/commits/1140b565c175cbdae114a39158f4639a373b04db).